### ISSUE TITLE
Use a f2b-specified table in nftables

### ISSUE THREAD

**Issue by holymonson** [fail2ban/fail2ban#1951]
In **iptables**, there are built-in tables (e.g. `filter`) and chains (e.g. `INPUT`) which controlling the flow in netfilter. So f2b majorly ban IPs by manipulating chain `INPUT`.

However, in **nftables**, there are **NO** built-in tables or chains. The filter point is specified by `type` and `hook` (and `priority`). see https://wiki.nftables.org/wiki-nftables/index.php/Configuring_chains
```
% nft add chain [<family>] <table-name> <chain-name> { type <type> hook <hook> priority <value> \; }
```

`table-name` and `chain-name` are irrelevant here. So instead of messing up with rules user-configured in `INPUT` like iptables, could we **setup a new f2b-specified table, only manipulate it and delete it when stopping**?

I test the below and it works for me. (Some variables are hardcoded because I am confused by loading priority.)
```
[Definition]

_nft_list =
_nft_get_handle_id =

actionstart = <nftables> add table <nftables_family> f2b-filter
              <nftables> add set <nftables_family> f2b-filter f2b-set \{ type <nftables_type>\; \}
              <nftables> add chain <nftables_family> f2b-filter f2b-input \{ type filter hook input priority 1 \; \}
              <nftables> add rule <nftables_family> f2b-filter f2b-input %(nftables_mode)s ip saddr f2b-set <blocktype>
actionstop = <nftables> delete table <nftables_family> f2b-filter
actioncheck = <nftables> list chain <nftables_family> f2b-filter f2b-input | grep -q '@f2b-set[ \t]'
actionban = <nftables> add element <nftables_family> f2b-filter f2b-set \{ <ip> \}
actionunban = <nftables> delete element <nftables_family> f2b-filter f2b-set \{ <ip> \}
```
We will get a table like 
```
table inet filter { // user configured table
	chain input {
		type filter hook input priority 0; policy accept;
                // [ user configured rules ]
	}
}

table inet f2b-filter {
	set f2b-set {
		type ipv4_addr
		elements = { 121.193.184.2, 5.188.10.180, 5.188.10.179, 31.207.47.50, 195.228.59.55, 193.201.224.241, 5.188.10.182}
	}

	chain f2b-input {
		type filter hook input priority 1; policy accept;
		tcp dport { ssh} ip saddr @f2b-set reject
	}
}
```

**Comment by sebres** [comment]
> could we setup a new f2b-specified table, only manipulate it and delete it when stopping?

Sure, but just think about multiple jails (with different ports, protocol etc.). If your "f2b-specified table" can do this, I've no problem with it. So welcome with the PR.

I assume that your current example will work only for exactly one jail, resp. in best case will ban tickets for all ports specified for this table (`f2b-filter`), if nft will accept such in-place extending with several ports for the same table.

Thus I think everywhere instead of fixed `f2b-filter` you should rather use an interpolation like `f2b-<name>` (which is currently a `<set_name>`) or similar, to decidedly differentiate between tables of different jails.

Additionally, I did not understood, how your approach should be better as current implementation of `nftables-multiport.conf`.

**Comment by holymonson** [comment]
> I assume that your current example will work only for exactly one jail, resp. in best case will ban tickets for all ports specified for this table (f2b-filter), if nft will accept such in-place extending with several ports for the same table.

Setting `port = http, https` or `port = 1-65535` will do. Note that it didn't hardcode `ssh` in the above `nftables-common.local` .
But for multiple jails, when stopping one we could delete the related rules instead of the whole table, so finally it may leave an empty table and empty chain. It won't do harm.

> Thus I think everywhere instead of fixed f2b-filter you should rather use an interpolation like f2b-<name> (which is currently a <set_name>) or similar, to decidedly differentiate between tables of different jails.

Seems no need. Jails can share a same table and same chain, they are differentiated by rules (and addr_set inside).

> Additionally, I did not understood, how your approach should be better as current implementation of nftables-multiport.conf.

1. nftables is considerd as the successor of iptables, it has several improvements. The current implementation follows the iptables logic so it doesn't show the advantages.
2. fail2ban handlers can be seperated from other user-configured rules, so the logic will be clearer and better coexistent with other software manipulating nftables too.
3. Table and chain are private for fail2ban, so we won't need a `nftables_table ` or `chain` parameter. So we won't have an issue like #1949 . 
4. ipv4 and ipv6 can be handled in the same table, so `nftables_family` is no need too. Only `address_family ` in use.

I will improve my `nftables-common.local` into a PR after this busy week, then you can see if you like it.

**Comment by sebres** [comment]
> Seems no need. Jails can share a same table and same chain, they are differentiated by rules (and addr_set inside).

You are sure, that then such configuration by the IP1 banned with jail-A (with TCP ports http,https...) and IP2 banned within jail-B (with UDP port X), are still allow to connect IP1 to port X and IP2 to port http?

If not, such PR will be accepted only for `nftables-allports` (but still protocol involved), resp. to something like `nftables-recidive` (for all ports / protocols).

> nftables is considerd as the successor of iptables, it has several improvements.

I don't see it so (because I'd not seen still any advantages (for fail2ban) compared for example with iptables+ipset);
> The current implementation follows the iptables logic so it doesn't show the advantages.

Could you explain it a bit more? I mean exactly "show the advantages".

> So we won't have an issue like #1949 .

#1949 is not an issue anymore (was fixed in b615a9854019589c87d68dd586f45c61f8fe886d).

> ipv4 and ipv6 can be handled in the same table, so nftables_family is no need too. 

If it's so, then this can be easy fixed in current implementation also (by removing corresponding conditional parameter)...

**Comment by holymonson** [comment]
> You are sure, that then such configuration by the IP1 banned with jail-A (with TCP ports http,https...) and IP2 banned within jail-B (with UDP port X), still allow to connect IP1 to port X and IP2 to port http?

Yes. Jails are implemented by rules, and rules are associated with different address-sets (`set_name = f2b-<name>`). So each jail will have its own address-set to ban.

> Could you explain it a bit more? I mean exactly "show the advantages".

In implement level, nftables has a simpler logic than iptables (ip6tables, arptables, etc.), it majorly runs in user space while iptables in kernel space, so it is easy to extend without modifying kernel. It needs less kAPI and its code is even smaller and clearer than all the Xtables. So I believe it will someday replace iptables like iptables to ipchains. some doc in http://wiki.nftables.org/ .

As for fail2ban, user MUST manually setup a table and a chain BEFORE starting fail2ban, so that rules can be inserted into it. (Unlike iptables, there is no predefined tables or chains in nftables).

Beside above and mentioned earlier, I have to admit that for now there is nothing nftables can do while the iptables-compatible way can't. Just like we can write **C** in `.cpp`, but why not use the mechanism provided by **C++**? Those prefer iptables will still use iptables, so is it better to let nftables works in a non-iptables-compatible way?

Of course, I am OK if you want less change, because "if the code works, don't touch it". There seems still few people using nftables in fail2ban now, we can wait and see if someone will bring up questions on nftables actions later.

**Comment by sebres** [comment]
> In implement level, nftables has a simpler logic than iptable...

I am familiar, you did not understood... I would just ask "show the advantages" **of your approach**, fail2ban-related, like in comparison to already implemented nftables-actions.

> As for fail2ban, user MUST manually setup a table and a chain BEFORE starting fail2ban

Is it really so? (I don't use nftables at all...).

> Of course, I am OK if you want less change, because "if the code works, don't touch it".

Nope. I'm never against a PR (especially if it is good justified). Just sometimes hard to find a tester for it :)

So we're waiting to see your PR :)

**Comment by ririsoft** [comment]
Hello,

I am migrating from iptables to nftables and would like to try this patch. Having a table dedicated to fail2ban and relying on the priority attribute to ensure f2b is processed first looks more "nftablesish" to me than the current implementation.

I have a question though, I am not a f2b expert sorry about that: I want to have both ipv4/ipv6 bans. This setup sets an "inet" table, but due to nftables sets limitation we must have 2 sets and rules for each protocol, this patch adds only one rule to the inet table.

What is your recommendation to enable both ipv4/ipv6 bans with such implementation ?
Thanks in advance !

Cheers.

**Comment by sebres** [comment]
> This setup sets an "inet" table ...

Do you mean the #1956? Because I see conditional section there "Init?family=inet6".
```INI
 [Init?family=inet6]
addr_family = ip6
addr_type = ipv6_addr
addr_set = addr6-set-<name>
```
So it looks good as regards the IPv6.
But please note the discussion with @albel727 in this PR.

**Comment by jrchamp** [comment]
Will be fixed by #2254

**Comment by sebres** [comment]
#2254 got merged now, many thanks to all involved.
